receiver/receivercreator/receiver.go

```diff
@@ -0,0 +1,154 @@
+// Copyright 2020, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package receivercreator
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/open-telemetry/opentelemetry-collector/component"
+	"github.com/open-telemetry/opentelemetry-collector/config"
+	"github.com/open-telemetry/opentelemetry-collector/config/configmodels"
+	"github.com/open-telemetry/opentelemetry-collector/consumer"
+	"github.com/open-telemetry/opentelemetry-collector/consumer/consumerdata"
+	"github.com/open-telemetry/opentelemetry-collector/oterr"
+	"github.com/spf13/viper"
+	"go.uber.org/zap"
+)
+
+var (
+	errNilNextConsumer = errors.New("nil nextConsumer")
+)
+
+var _ component.MetricsReceiver = (*receiverCreator)(nil)
+
+// receiverCreator implements consumer.MetricsConsumer.
+type receiverCreator struct {
+	nextConsumer consumer.MetricsConsumerOld
+	logger       *zap.Logger
+	cfg          *Config
+	receivers    []component.Receiver
+}
+
+// New creates the receiver_creator with the given parameters.
+func New(logger *zap.Logger, nextConsumer consumer.MetricsConsumerOld, cfg *Config) (component.MetricsReceiver, error) {
+	if nextConsumer == nil {
+		return nil, errNilNextConsumer
+	}
+
+	r := &receiverCreator{
+		logger:       logger,
+		nextConsumer: nextConsumer,
+		cfg:          cfg,
+	}
+	return r, nil
+}
+
+// loadRuntimeReceiverConfig loads the given subreceiverConfig merged with config values
+// that may have been discovered at runtime.
+func (dr *receiverCreator) loadRuntimeReceiverConfig(factory component.ReceiverFactoryOld, staticSubConfig *subreceiverConfig, subConfigFromEnv map[string]interface{}) (configmodels.Receiver, error) {
+	// Load config under <receiver>/<id> since loadReceiver and CustomUnmarshaler expects this structure.
+	viperConfig := viper.New()
+	viperConfig.Set(staticSubConfig.fullName, map[string]interface{}{})
+	subreceiverConfig := viperConfig.Sub(staticSubConfig.fullName)
```

Comment:
Use this:
https://godoc.org/github.com/open-telemetry/opentelemetry-collector/config#NewViper

See also some bugs in Sub https://github.com/open-telemetry/opentelemetry-collector/pull/707

Reply:
@bogdandrutu should `viperSub` be exported or just copy the function into here?

Reply:
You can do a PR to export that temporary.

Reply:
I updated to just not use sub as it didn't provide much value.